core/shared/src/main/scala/scalaz/zio/IO.scala

```diff
@@ -998,7 +993,16 @@ object IO extends Serializable {
    * Imports an asynchronous effect into a pure `IO` value. This formulation is
    * necessary when the effect is itself expressed in terms of `IO`.
    */
-  final def asyncPure[E, A](register: (Callback[E, A]) => IO[E, Unit]): IO[E, A] = new AsyncIOEffect(register)
+  final def asyncPure[E, A](register: (Callback[E, A]) => IO[E, Unit]): IO[E, A] =
+    for {
+      p   <- Promise.make[E, A]
+      ref <- Ref[Fiber[E, Any]](Fiber.unit)
+      a <- (for {
+            f <- register(p.unsafeDone(_)).fork.peek(ref.set(_)).uninterruptibly
+            _ <- f.join
```

Comment:
Is the `join` necessary here? in particular, the `IO` returned by `register` may not complete until after register's callback is invoked

Reply:
Not sure if it's strictly necessary, but the join will help ensure that if the `register` is interrupted, the parent will be interrupted too.